File: HarryPotterApp/app/src/main/java/com/antelo97/harrypotterapp/data/repository/CharacterRepository.kt
package com.antelo97.harrypotterapp.data.repository

import com.antelo97.harrypotterapp.data.database.dao.CharacterDao
import com.antelo97.harrypotterapp.data.database.model_entity.CharacterEntity
import com.antelo97.harrypotterapp.data.database.model_entity.toDatabase
import com.antelo97.harrypotterapp.data.network.api.CharacterService
import com.antelo97.harrypotterapp.domain.model.toDomain
import javax.inject.Inject

class CharacterRepository @Inject constructor(
    private val api: CharacterService,
    private val dao: CharacterDao
) {

    suspend fun getCharactersFromDatabase(): List<com.antelo97.harrypotterapp.domain.model.Character> {
        val query: List<CharacterEntity> = dao.getCharacters()
        return query.map { it.toDomain() }
    }

    suspend fun insertCharactersAndWands() {
        val characters: List<CharacterEntity> = api.getAllCharacters().map { it.toDatabase() }

        deleteCharacters()
        dao.insertCharactersAndWands(characters)
    }

    suspend fun deleteCharacters() {
        dao.deleteCharacters()
    }

    suspend fun getCharactersByName(searchQuery: String): List<com.antelo97.harrypotterapp.domain.model.Character> {
        return dao.getCharactersByName(searchQuery).map { it.toDomain() }
    }
}

File: flutter_projects/harry_potter_app/lib/domain/repository/character_repository.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:harry_potter_app/data/cloud_firebase_db/cloud_constants.dart';
import 'package:harry_potter_app/data/network/api/character_service.dart';
import 'package:harry_potter_app/domain/model/character.dart';

class CharacterRepository {
  final api = CharacterService();
  final charactersCollection =
      FirebaseFirestore.instance.collection('characters');

  static final CharacterRepository _shared =
      CharacterRepository._sharedInstance();
  CharacterRepository._sharedInstance(); // Private constructor
  factory CharacterRepository() => _shared;

  Future<List<Character>> loadCharactersFromApi() async {
    final characters = await getCharactersFromCloudFirebase();

    if (characters.isEmpty) {
      insertCharacters();
      return getCharactersFromCloudFirebase();
    } else {
      // deleteCharacters();
      // insertCharacters();
      return characters;
    }
  }

  Future<List<Character>> getCharactersFromCloudFirebase() async {
    final snapshot = await charactersCollection.get();
    if (snapshot.docs.length >= 2) {
      return snapshot.docs.map((doc) => Character.fromDocument(doc)).toList();
    } else {
      return [];
    }
  }

  Future<void> insertCharacters() async {
    final charactersResponse = await api.getCharacters();
    final characters = charactersResponse!
        .map((characterResponse) => Character.fromResponse(characterResponse))
        .toList();
    final batch = FirebaseFirestore.instance.batch();

    for (var character in characters) {
      final idDoc = charactersCollection.doc();
      batch.set(
        idDoc,
        character.toMap(),
      );
    }

    await batch.commit();
    deleteDefaultEmptyDocument();
  }

  Future<void> deleteCharacters() async {
    final snapshot = await charactersCollection.get();
    final batch = FirebaseFirestore.instance.batch();

    for (var doc in snapshot.docs) {
      batch.delete(doc.reference);
    }

    await batch.commit();
  }

  Future<List<Character>> searchCharactersByName(String searchQuery) async {
    return await charactersCollection
        .where(nameFieldName, isGreaterThanOrEqualTo: searchQuery.toLowerCase())
        .where(nameFieldName,
            isLessThanOrEqualTo: '${searchQuery.toLowerCase()}\uf8ff')
        .get()
        .then((snapshot) =>
            snapshot.docs.map((doc) => Character.fromDocument(doc)).toList());
  }

  Future<void> updateCharacter(Character character) async {
    try {
      await charactersCollection.doc(character.idDocument).update({
        idApiCharacterFieldName: character.idApiCharacter,
        nameFieldName: character.name,
        speciesFieldName: character.species,
        genderFieldName: character.gender,
        houseFieldName: character.house,
        yearOfBirthFieldName: character.yearOfBirth,
        isWizardFieldName: character.isWizard,
        ancestryFieldName: character.ancestry,
        wandFieldName: character.wand,
        patronusFieldName: character.patronus,
        isHogwartsStudentFieldName: character.isHogwartsStudent,
        isHogwartsStaffFieldName: character.isHogwartsStaff,
        actorFieldName: character.actor,
        isAliveFieldName: character.isAlive,
        imageUrlFieldName: character.imageUrl,
      });
    } catch (e) {
      // CouldNotUpdateNoteException();
    }
  }

  List<Character> sortCharactersByNameAsc(List<Character> characters) {
    characters.sort((oneCharacter, anotherCharacter) =>
        oneCharacter.name.compareTo(anotherCharacter.name));
    return characters;
  }

  List<Character> sortCharactersByNameDesc(List<Character> characters) {
    characters.sort((oneCharacter, anotherCharacter) =>
        anotherCharacter.name.compareTo(oneCharacter.name));
    return characters;
  }

  void deleteDefaultEmptyDocument() async {
    final snapshot = await charactersCollection.get();

    for (var doc in snapshot.docs) {
      final data = doc.data();

      if (data.isEmpty) {
        await doc.reference.delete();
      }
    }
  }
}
